### ISSUE TITLE
Is there a way to link the features of model A to the output of model B?

### ISSUE THREAD

**Issue by leoentersthevoid** [SeldonIO/alibi#380]
First of all, thanks for all the effort you put into creating a consistent framework.

Now, suppose your data consist of many different frames, all of which come from different videos.
You use IsolationForest (**model A**) and you end up getting some predictions from the set `{-1, 1}`.
Now your problem is that your labels are at the video level, not the frame level, so you end up fitting a Gaussian (**model B**) on the means of all the predictions for a given video and get predictions from the set `{0, 1}` (your class, ie. anomaly or not), based on how many standard deviations from the mean is the mean of model A outputs.

Is there a way to explain model B outputs based on the initial set of features model A used?

**Comment by jklaise** [comment]
@leoentersthevoid if I understand correctly then you are using model A predictions as features for model B? It should be possible to do explanations by defining your black box model as a function taking in the original features (frames) and outputing the predictions from model B, essentially wrapping the two models into one prediction function. Does that make sense?

**Comment by leoentersthevoid** [comment]
@jklaise, that was my initial thought as well. however, as far as I experimented, I've seen that SHAP can't handle a dimensionality reduction in the output space, meaning if

model_A_in: (a, b)
model_A_out: (a, )
model_B_in: (k, )
model_B_out: (k, )

where k < a

`shap.KernelExplainer` (or any other explainer) didn't help. 

The only way it would work is if `k = a`, which I can't have in my setup.

**Comment by jklaise** [comment]
@leoentersthevoid not sure I follow your example, what exactly are `a, b, k` and how are you combining this models into one predictor?

**Comment by leoentersthevoid** [comment]
class CustomIsolationForest:
    def __init__(self, std_threshold=3):
        self.if_forest = IsolationForest(random_state=31337, contamination="auto", n_jobs=-1)
        self.mean, self.std = 0, 0
        self.std_threshold_ = std_threshold
        self.is_fitted = False
    
    def predict(self, data, idxs=None):
        if not self.is_fitted:
            raise ValueError("Fit the model first!")
        if idxs is not None:
            preds = self.if_forest.predict(data)
            outliers = []
            k = 0
            for i in idxs:
                outliers.append(self.__aggregate(preds[k:k+i]))
                k += i
            return np.array(outliers)
        raise ValueError('Missing indexes!')


    def fit(self, data, idxs=None):
        if idxs is not None:
            self.if_forest.fit(data) # this is model A; data will have a shape of (a, b), it's a matrix
            preds = self.if_forest.predict(data) # predictions will have a shape (a, ), it's a vector of values from the set {-1, 1}
            means = []
            k = 0
            for i in idxs:
                means.append(preds[k:k+i].mean()) 
                k += i
            self.mean, self.std = fit_a_distribution_function.fit(means) # this is model B; means will have a shape of (k, ), it's a vector of values from the interval [-1, 1]
            self.is_fitted = True
        raise ValueError('Missing indexes!')

    def __aggregate(self, data):
        return int(data.mean() < self.mean - self.std_threshold_ * self.std) # 0 or 1 


X_train, train_idxs, X_test, test_idxs, y_test = function_that_returns_the_datasets(*paths)

model = CustomIsolationForest()
model.fit(X_train, train_idxs)
preds = model.predict(X_test, test_idxs) # will have the same dimension as y_test and values from the set {0, 1}
whatever_metric(preds, y_test)
```
So, as you can see from above, you go from an input that has a shape of `(a, b)` to an output that has a shape of `(k, )`. 

How do you make the explanation?

**Comment by jklaise** [comment]
Just to clarify, you are putting in `a` datapoints with `b` features each and the output is a vector of length `k`? I'm not sure it makes much sense to talk about explanations in this scenario because the output does not map one-to-one to the input... Maybe it's worth expanding on what this k-dimensional output represents?

**Comment by leoentersthevoid** [comment]
No... 

`a` is the length of the input (ie. `shape[0]`, not `shape[1]` which would be the features)
`b` is the length of the output of the first model
`k` is the length of the output of the second model

But nonetheless, I reached your conclusion as well. 

I went explaining the first model independently of the second model, and I'm planning to do so for the second model as well. 
Then will look at whether or not there is a way to link the explanations.

Thanks for your time!
